File: packages/typescript/src/pdf/wht-credit-note.ts
// ─── WHT Credit Note PDF (Layer 2) ───────────────────────────────────────────
// Generates WHT credit note PDFs using pdfkit — no headless browser required.
// Returns a Buffer, performs no file I/O. Serverless-compatible.

import PDFDocument from 'pdfkit';

// ─── Types ───────────────────────────────────────────────────────────────────

export interface WhtCreditNoteParty {
  name: string;
  tin: string;
  address: string;
}

export interface WhtCreditNoteOptions {
  deductor: WhtCreditNoteParty;
  beneficiary: WhtCreditNoteParty;
  grossAmount: number;
  whtRate: number;
  whtAmount: number;
  netPayment: number;
  serviceDescription: string;
  paymentDate: string;
  creditNoteNumber: string;
  remittanceReceiptNumber?: string;
  legalBasis?: string;
}

export interface WhtCreditNote {
  deductor: WhtCreditNoteParty;
  beneficiary: WhtCreditNoteParty;
  grossAmount: number;
  whtRate: number;
  whtAmount: number;
  netPayment: number;
  serviceDescription: string;
  paymentDate: string;
  creditNoteNumber: string;
  remittanceReceiptNumber?: string;
  legalBasis: string;
}

// ─── Formatting Helpers ──────────────────────────────────────────────────────

function formatCurrency(amount: number): string {
  const fixed = Math.abs(amount).toFixed(2);
  const [whole, decimal] = fixed.split('.');
  const withCommas = whole.replace(/\B(?=(\d{3})+(?!\d))/g, ',');
  const sign = amount < 0 ? '-' : '';
  return `${sign}\u20A6${withCommas}.${decimal}`;
}

function formatDate(iso: string): string {
  const parts = iso.split('-');
  if (parts.length !== 3) return iso;
  return `${parts[2]}/${parts[1]}/${parts[0]}`;
}


// ─── Layout Constants ────────────────────────────────────────────────────────

const PAGE_MARGIN = 50;
const PAGE_WIDTH = 595.28;
const CONTENT_WIDTH = PAGE_WIDTH - PAGE_MARGIN * 2;
const FONT_BOLD = 'Helvetica-Bold';
const FONT_REGULAR = 'Helvetica';
const COLOR_PRIMARY = '#1a1a2e';
const COLOR_SECONDARY = '#555555';
const COLOR_LINE = '#cccccc';

// ─── Create ──────────────────────────────────────────────────────────────────

/**
 * Create a WhtCreditNote from options.
 * Validates and populates all required fields including legal basis.
 */
export function create(options: WhtCreditNoteOptions): WhtCreditNote {
  return {
    deductor: options.deductor,
    beneficiary: options.beneficiary,
    grossAmount: options.grossAmount,
    whtRate: options.whtRate,
    whtAmount: options.whtAmount,
    netPayment: options.netPayment,
    serviceDescription: options.serviceDescription,
    paymentDate: options.paymentDate,
    creditNoteNumber: options.creditNoteNumber,
    remittanceReceiptNumber: options.remittanceReceiptNumber,
    legalBasis: options.legalBasis ?? 'WHT Regulations 2024; NTA 2025 Part IV',
  };
}

// ─── PDF Generation ──────────────────────────────────────────────────────────

/**
 * Generate a WHT Credit Note PDF as a Buffer.
 * Renders: deductor TIN, beneficiary TIN, gross amount, WHT rate,
 * WHT amount, net payment, remittance receipt number, legal basis.
 */
export function toPDF(note: WhtCreditNote): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    const doc = new PDFDocument({
      size: 'A4',
      margin: PAGE_MARGIN,
      info: {
        Title: `WHT Credit Note ${note.creditNoteNumber}`,
        Author: note.deductor.name,
      },
    });

    const chunks: Uint8Array[] = [];
    doc.on('data', (chunk: Uint8Array) => chunks.push(chunk));
    doc.on('end', () => resolve(Buffer.concat(chunks)));
    doc.on('error', (err: Error) => reject(err));

    // ── Header ──
    doc
      .font(FONT_BOLD).fontSize(14).fillColor(COLOR_PRIMARY)
      .text('WHT CREDIT NOTE', PAGE_MARGIN, PAGE_MARGIN, { align: 'center', width: CONTENT_WIDTH });

    doc.moveDown(0.5);
    doc.font(FONT_REGULAR).fontSize(9).fillColor(COLOR_SECONDARY)
      .text(`Credit Note #: ${note.creditNoteNumber}`, { align: 'center', width: CONTENT_WIDTH });
    doc.text(`Payment Date: ${formatDate(note.paymentDate)}`, { align: 'center', width: CONTENT_WIDTH });

    doc.moveDown(1);

    // ── Deductor Details ──
    doc.font(FONT_BOLD).fontSize(10).fillColor(COLOR_PRIMARY).text('Deductor (Payer)');
    doc.font(FONT_REGULAR).fontSize(9).fillColor(COLOR_SECONDARY);
    doc.text(`Name: ${note.deductor.name}`);
    doc.text(`TIN: ${note.deductor.tin}`);
    doc.text(`Address: ${note.deductor.address}`);

    doc.moveDown(0.8);

    // ── Beneficiary Details ──
    doc.font(FONT_BOLD).fontSize(10).fillColor(COLOR_PRIMARY).text('Beneficiary (Payee)');
    doc.font(FONT_REGULAR).fontSize(9).fillColor(COLOR_SECONDARY);
    doc.text(`Name: ${note.beneficiary.name}`);
    doc.text(`TIN: ${note.beneficiary.tin}`);
    doc.text(`Address: ${note.beneficiary.address}`);

    doc.moveDown(1);

    // ── Payment Details ──
    doc.font(FONT_BOLD).fontSize(10).fillColor(COLOR_PRIMARY).text('Payment Details');
    doc.moveDown(0.3);

    const detailsX = PAGE_MARGIN;
    const valueX = PAGE_MARGIN + 200;

    const rows: [string, string][] = [
      ['Service Description:', note.serviceDescription],
      ['Gross Amount:', formatCurrency(note.grossAmount)],
      ['WHT Rate:', `${(note.whtRate * 100).toFixed(1)}%`],
      ['WHT Amount:', formatCurrency(note.whtAmount)],
      ['Net Payment:', formatCurrency(note.netPayment)],
    ];

    if (note.remittanceReceiptNumber) {
      rows.push(['NRS Remittance Receipt #:', note.remittanceReceiptNumber]);
    }

    doc.font(FONT_REGULAR).fontSize(9).fillColor(COLOR_SECONDARY);
    for (const [label, value] of rows) {
      const y = doc.y;
      doc.font(FONT_BOLD).text(label, detailsX, y, { continued: false });
      doc.font(FONT_REGULAR).text(value, valueX, y);
    }

    doc.moveDown(1.5);

    // ── Legal Basis ──
    doc.moveTo(PAGE_MARGIN, doc.y).lineTo(PAGE_MARGIN + CONTENT_WIDTH, doc.y)
      .strokeColor(COLOR_LINE).lineWidth(0.5).stroke();
    doc.y += 8;

    doc.font(FONT_REGULAR).fontSize(7).fillColor(COLOR_SECONDARY)
      .text(`Legal Basis: ${note.legalBasis}`, PAGE_MARGIN, doc.y, { width: CONTENT_WIDTH });

    doc.moveDown(0.5);
    doc.text(
      'This credit note certifies that Withholding Tax has been deducted and will be remitted to the Nigeria Revenue Service (NRS) on behalf of the beneficiary.',
      PAGE_MARGIN, doc.y, { width: CONTENT_WIDTH },
    );

    doc.end();
  });
}

File: packages/python/src/ngtaxkit/pdf/wht_credit_note.py
"""WHT Credit Note PDF generation using fpdf2."""

from __future__ import annotations

from typing import Any

from . import _require_fpdf2
from .helpers import (
    COLOR_HEADER_BG, COLOR_LINE, COLOR_PRIMARY, COLOR_SECONDARY, COLOR_WHITE,
    CONTENT_WIDTH, PAGE_MARGIN, format_currency, format_date,
)


def _cur(amount: float) -> str:
    """Format currency for PDF output (latin-1 safe)."""
    return format_currency(amount, pdf_safe=True)


def create(
    *,
    deductor: dict[str, Any],
    beneficiary: dict[str, Any],
    gross_amount: float,
    wht_rate: float,
    wht_amount: float,
    net_payment: float,
    service_description: str,
    payment_date: str,
    credit_note_number: str,
    remittance_receipt_number: str | None = None,
    legal_basis: str | None = None,
) -> dict[str, Any]:
    """Build a WHT Credit Note data dict."""
    return {
        "deductor": deductor,
        "beneficiary": beneficiary,
        "gross_amount": gross_amount,
        "wht_rate": wht_rate,
        "wht_amount": wht_amount,
        "net_payment": net_payment,
        "service_description": service_description,
        "payment_date": payment_date,
        "credit_note_number": credit_note_number,
        "remittance_receipt_number": remittance_receipt_number,
        "legal_basis": legal_basis or "WHT Regulations 2024; NTA 2025 Part IV",
    }


def to_pdf(note: dict[str, Any]) -> bytes:
    """Generate a WHT Credit Note PDF as bytes."""
    fpdf = _require_fpdf2()
    pdf = fpdf.FPDF(orientation="P", unit="mm", format="A4")
    pdf.set_auto_page_break(auto=True, margin=20)
    pdf.add_page()

    # -- WHT CREDIT NOTE banner --
    pdf.set_fill_color(*COLOR_PRIMARY)
    pdf.set_text_color(*COLOR_WHITE)
    pdf.set_font("Helvetica", "B", 14)
    pdf.cell(CONTENT_WIDTH, 10, "WHT CREDIT NOTE", fill=True, align="C",
             new_x="LMARGIN", new_y="NEXT")
    pdf.ln(4)

    # -- Credit note number and payment date --
    pdf.set_font("Helvetica", "", 9)
    pdf.set_text_color(*COLOR_SECONDARY)
    pdf.cell(0, 4, f"Credit Note #: {note['credit_note_number']}",
             new_x="LMARGIN", new_y="NEXT")
    pdf.cell(0, 4, f"Payment Date: {format_date(note['payment_date'])}",
             new_x="LMARGIN", new_y="NEXT")
    pdf.ln(6)

    # -- Deductor (Payer) section --
    pdf.set_font("Helvetica", "B", 10)
    pdf.set_text_color(*COLOR_PRIMARY)
    pdf.cell(0, 5, "Deductor (Payer):", new_x="LMARGIN", new_y="NEXT")
    pdf.set_font("Helvetica", "", 9)
    pdf.set_text_color(*COLOR_SECONDARY)
    pdf.cell(0, 4, note["deductor"]["name"], new_x="LMARGIN", new_y="NEXT")
    pdf.cell(0, 4, f"TIN: {note['deductor']['tin']}",
             new_x="LMARGIN", new_y="NEXT")
    pdf.cell(0, 4, note["deductor"]["address"], new_x="LMARGIN", new_y="NEXT")
    pdf.ln(4)

    # -- Beneficiary (Payee) section --
    pdf.set_font("Helvetica", "B", 10)
    pdf.set_text_color(*COLOR_PRIMARY)
    pdf.cell(0, 5, "Beneficiary (Payee):", new_x="LMARGIN", new_y="NEXT")
    pdf.set_font("Helvetica", "", 9)
    pdf.set_text_color(*COLOR_SECONDARY)
    pdf.cell(0, 4, note["beneficiary"]["name"],
             new_x="LMARGIN", new_y="NEXT")
    pdf.cell(0, 4, f"TIN: {note['beneficiary']['tin']}",
             new_x="LMARGIN", new_y="NEXT")
    pdf.cell(0, 4, note["beneficiary"]["address"],
             new_x="LMARGIN", new_y="NEXT")
    pdf.ln(6)

    # -- Payment Details table --
    pdf.set_font("Helvetica", "B", 10)
    pdf.set_text_color(*COLOR_PRIMARY)
    pdf.cell(0, 5, "Payment Details", new_x="LMARGIN", new_y="NEXT")
    pdf.ln(2)

    col_label = 90
    col_value = 90

    pdf.set_fill_color(*COLOR_HEADER_BG)
    pdf.set_font("Helvetica", "B", 8)
    pdf.set_text_color(*COLOR_PRIMARY)
    pdf.cell(col_label, 6, "Item", fill=True, align="L")
    pdf.cell(col_value, 6, "Details", fill=True, align="R")
    pdf.ln()

    pdf.set_font("Helvetica", "", 9)
    pdf.set_text_color(*COLOR_SECONDARY)

    rows = [
        ("Service Description", note["service_description"]),
        ("Gross Amount", _cur(note["gross_amount"])),
        ("WHT Rate", f"{note['wht_rate'] * 100:.1f}%"),
        ("WHT Amount", _cur(note["wht_amount"])),
        ("Net Payment", _cur(note["net_payment"])),
    ]
    if note.get("remittance_receipt_number"):
        rows.append(("Remittance Receipt #", note["remittance_receipt_number"]))

    for label, value in rows:
        pdf.cell(col_label, 5, label, align="L")
        pdf.cell(col_value, 5, value, align="R")
        pdf.ln()
        pdf.set_draw_color(*COLOR_LINE)
        pdf.line(pdf.l_margin, pdf.get_y(),
                 pdf.l_margin + CONTENT_WIDTH, pdf.get_y())
        pdf.ln(1)

    pdf.ln(6)

    # -- Legal basis footer --
    pdf.set_draw_color(*COLOR_LINE)
    pdf.line(pdf.l_margin, pdf.get_y(),
             pdf.l_margin + CONTENT_WIDTH, pdf.get_y())
    pdf.ln(3)
    pdf.set_font("Helvetica", "", 7)
    pdf.set_text_color(*COLOR_SECONDARY)
    pdf.multi_cell(CONTENT_WIDTH, 3,
                   f"Legal Basis: {note['legal_basis']}")
    pdf.ln(2)
    pdf.cell(CONTENT_WIDTH, 3,
             f"Generated by ngtaxkit -- WHT Credit Note {note['credit_note_number']}",
             align="C")

    return pdf.output()
